### DISCUSSION TITLE
Recommended way to implement plugin configuration

### DISCUSSION THREAD

**Discussion by copiousfreetime** [puma/puma#3173] [Q&A] [answered]
I'm writing a plugin that needs to be configured, and I see 3 different ways of doing this configuration now, and few that would require an update to puma itself,  and I'd like to get feedback on ways to do this.

Given a Plugin like the following that needs a parameter configured:

```ruby
::Puma::Plugin.create do
  def config(config)
  # config items -- maybe update some other item from the config
  end

 def start(launcher)
   thing_parameter = # get the parameter from somewhere
   @thing = Thing.new(thing_parameter)
   in_background do
       # Thing that needs a config item set
      @thing.do_your_thing
   end
 end
end
```

# Existing methods
## Environment varaibles

This seems to be the most commonly used method of the plugins I've reviewed.

```ruby
# puma/plugin/thing.rb
 def start(launcher)
   thing_parameter = ENV.fetch('THING_PARAM', Thing::DEFAULT_PARAM)
  # ...
 end
```

## Configuration object within the plugin itself, or elsewhere

This, or similar method is used by [puma-plugin-telemetry](https://github.com/babbel/puma-plugin-telemetry/tree/main)

In the plugin

```ruby
::Puma::Plugin.create do 
  class << self
    attr_writer :thing_param
    def thing_param
        @thing_param ||= Thing::DEFAULT_PARAM
    end
  end
  # rest of plugin from above
end
```

Or global config

```ruby
class Thing
  class << self
    attr_writer :thing_param
    def thing_param
      @thing_param ||= Thing::DEFAULT_PARAM
    end
  end
end
```

Or some global `Config` object that is created when the `puma.rb` file is parsed.

And then in the puma.rb file

```ruby
plugin :thing
::Thing.thing_param = :foo
# or
::Puma::Plugins::Thing.thing_param = :foo
# or
::Thing.configure do |config|
  config.thing_param = :foo
end
```

## Extend the `::Puma::DSL` class

This is the method used by the [puma-redeploy plugin](https://github.com/tbeauvais/puma-redeploy/blob/master/lib/puma/redeploy/dsl.rb).

In the plugin codebase somewhere - reopen the `Puma::DSL` class

```ruby
module Puma
  class DSL
     def thing_param(value)
      @options[:thing_param] = value
    end
  end
end
```

And then in `puma.rb`

```ruby
require 'thing'
thing_param :foo
plugin :thing
```

And in the plugin
```ruby
# puma/plugin/thing.rb

  def start(launcher)
    thing_param = launcher.config.thing_param
    #...
  end
```

# Additional Ideas

These are a few ideas for doing plugin configuration, I have not attempted any of them, so this is just potential examples

## Allow parameters to be passed to the plugins initializer, or yielded

Currently, when a plugin is instantiated:

https://github.com/puma/puma/blob/763d1a1a971d9bb17fc1478859d626a539c29970/lib/puma/plugin.rb#L11-L16

No parameters are passed to the the plugins `initialize` method.  If we could pass in parameters that might make things easier, either directly, or yield the instance

```ruby
# puma.rb
plugin :thing, { thing_parameter: value }

# or yield the instance
plugin :thing do |thing_instance|
  thing_instance.thing_parameter = value
end
```
In either of these cases, we could end up with an instance variable, or something in the plugin instance running and it could do:

```ruby
# puma/plugin/thing.rb
#...
  def start(launcher)
   @thing = Thing.new(self.thing_parameter)
   in_background do
       # Thing that needs a config item set
      @thing.do_your_thing
   end
```

## Known plugin namespace in the puma configuration

This idea would not change the plugin instantiation, and would allow for items from the `puma.rb` file to be reference by the instantiated plugin through the `Launcher` instance.

```ruby
# puma.rb
plugin :thing
plugin_config[:thing].thing_param = :foo`
```

It could be that passing in params to the `plugin` method would facilitate this too, parameters passed

```ruby
plugin :thing, { thing_param: :foo } # the hash would be made available in the launcher config
```

```ruby
# puma/plugin/thing.rb

def config(config)
  config.plugin_config[:thing]  # => { thing_param: :foo }
end

def start(launcher)
   thing_param = launcher.config.plugin_config[:thing] # => { thing_param: :foo }
   # rest of method
end
```

# Thoughts

My personal thoughts are that puma should provide some mechanism that allows plugin specific configuration to flow from the `puma.rb` file into the plugin instances. 

After writing all of this, I'm currently learning towards the plugin specific configuration namespace that is manage by puma and is used in the `puma.rb` as a 2nd parameter to `plugin`

```ruby
plugin :thing, { thing_param: :foo }
```

I'm completely open to any and all recommendations

**Comment by nateberkopec** [comment]
1. Env variables isn't great, it means config isn't checked in to the application repository probably (or it's difficult to checkin, using .env or something like this)
2. Extending Puma::DSL is dangerous because Puma might override your method someday.
3. ```::Thing.configure``` in your `puma.rb` file is probably the best option available today.

So, what do we gain from adding `plugin :thing, { thing_param: :foo }`? Nicer syntax, but is functionality improved?

**Reply by copiousfreetime** [reply]
Thanks for the feedback @nateberkopec. I definitely agree with you on (1). Although, at the moment, I prefer (2) just slightly (and its not by much) more than (3) because (2) has a more consistent configuration syntax, since puma itself is configured with the dsl methods.

```ruby
plugin :thing
thing_config_item_1 'one'
thing_config_item_2 'two'
```
vs

```ruby
plugin :thing
Thing.configure do |config|
  config.item_1 = 'one'
  config.item_2 = 'two'
end

# or
plugin :thing
::Thing.config.item_1 = 'one'
::Thing.config.item_2 = 'two'
```

Both of those later ones just don't sit right with me. different styles what don't match the configuration pattern. Maybe its too much of a nitpick.

 🤷 Odds of a puma configuration namespace collision with another plugin vs. maintaining a global configuration object  🤷.

As for adding in something like `plugin :thing, { thing_param: :foo }`  or `plugin :thing do |thing_plugin_instance| ... end`; I see 2 benefits:

1. From an end-user / developer perspecitve, Yes, it is a nicer syntax, which is I think is a functional improvement as it may mitigate confusion.
2. From a plugin developer perspective, we have a consistent interface for both configuration and for operation of the plugin. I don't have to write a whole other set of code to store and fetch configuration, I can just write:

```ruby
Puma::Plugin.create do
  def initialize(**args)
   # get the args passed in from the config 
  end
  def config(conf); 
  # or get the params from conf.plugin_config[:thing]
  end
  def start(launcher) ; end
end
```

or 

```ruby
Puma::Plugin.create do
    attr_accessor :config_item_1
    attr_accessor :config_item_1
end
```
and know that in the puma config I can just set the the attributes in a block attached to `plugin :thing...`

I would be happy to work  a PR that would add this type of functionality, it could probably support both of these options if we wanted and resolve ambiguity in writing puma plugins. Would a PR adding this be considered?

In the meantime, I'll probably go with extending the DSL approach.

**Reply by nateberkopec** [reply, marked as answer]
Re: 2, if you simply prepend your DSL method with something at least reasonably unique (like your plugin name, as you did in your example) then it's highly unlikely we'll ever have a conflict.

I think Puma is too mature of a project to add things like the originally proposed interface... 2 and 3 are for me "good enough"

**Reply by copiousfreetime** [reply]
Thanks for the consideration @nateberkopec. I'm 100% behind "Good Enough". And yes, prefixing DSL methods with the plugin name is what we've decided to go with.
